File: OC/显示层动画/WGOCAnimationPrimary/WGOCAnimationPrimary/关键帧动画/转盘/WGTurnplateKeyViewController.m
//
//  WGTurnplateKeyViewController.m
//  WGOCAnimationPrimary
//
//  Created by wanggang on 2019/5/21.
//  Copyright © 2019 bozhong. All rights reserved.
//

#import "WGTurnplateKeyViewController.h"

@interface WGTurnplateKeyViewController ()

@property (nonatomic, strong) UIImageView *imgView;
@property (nonatomic, assign) NSInteger index;

@end

@implementation WGTurnplateKeyViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.title = @"转盘";
    self.view.backgroundColor = UIColor.whiteColor;
    
    self.imgView = [[UIImageView alloc] initWithFrame:CGRectMake((SCREEN_WIDTH-300)/2, (SCREEN_HEIGHT-300)/2, 300, 300)];
    self.imgView.image = [UIImage imageNamed:@"turnPlate"];
    [self.view addSubview:self.imgView];
}

- (void)viewWillAppear:(BOOL)animated{
    [super viewWillAppear:animated];
    
    [UIView animateKeyframesWithDuration:0.2 delay:0 options:UIViewKeyframeAnimationOptionCalculationModeCubic animations:^{
        
        [UIView addKeyframeWithRelativeStartTime:0 relativeDuration:0.25 animations:^{
            
            self.index += 1;
            self.imgView.transform = CGAffineTransformMakeRotation(M_PI_2*self.index);
        }];
        
        [UIView addKeyframeWithRelativeStartTime:0.25 relativeDuration:0.25 animations:^{
            
            self.index += 1;
            self.imgView.transform = CGAffineTransformMakeRotation(M_PI_2*self.index);
        }];
        
        [UIView addKeyframeWithRelativeStartTime:0.5 relativeDuration:0.25 animations:^{
            
            self.index += 1;
            self.imgView.transform = CGAffineTransformMakeRotation(M_PI_2*self.index);
        }];
        
        [UIView addKeyframeWithRelativeStartTime:0.75 relativeDuration:0.25 animations:^{
            
            self.index += 1;
            self.imgView.transform = CGAffineTransformMakeRotation(M_PI_2*self.index);
        }];
        
    } completion:nil];
}

@end

File: Swift/显示层初级动画/WGAnimationPrimary/WGAnimationPrimary/关键帧动画/WGTurnplateKeyViewController.swift
//
//  WGTurnplateKeyViewController.swift
//  WGAnimationPrimary
//
//  Created by wanggang on 2019/4/22.
//  Copyright © 2019 bozhong. All rights reserved.
//

import UIKit

class WGTurnplateKeyViewController: UIViewController {

    var imgView: UIImageView?
    var index:CGFloat = 1.0
    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        imgView = UIImageView()
        imgView?.bounds = CGRect(x: 0, y: 0, width: 300, height: 300)
        imgView?.center = self.view.center
        imgView?.image = UIImage(named: "turnPlate")
        imgView?.backgroundColor = UIColor.white
        self.view.backgroundColor = UIColor.white
        self.view.addSubview(imgView!)
        
        startAnimation()
    }
    
    func startAnimation() {
        
        UIView.animateKeyframes(withDuration: 0.2, delay: 0, options: UIView.KeyframeAnimationOptions(), animations: {
            
            UIView.addKeyframe(withRelativeStartTime: 0, relativeDuration: 1/4, animations: {
                
                self.index += 1;
                self.imgView?.transform = CGAffineTransform(rotationAngle: .pi/2*self.index)
            })
            
            UIView.addKeyframe(withRelativeStartTime: 1/4, relativeDuration: 1/4, animations: {
                
                self.index += 1;
                self.imgView?.transform = CGAffineTransform(rotationAngle: .pi/2*self.index)
            })
            
            UIView.addKeyframe(withRelativeStartTime: 2/4, relativeDuration: 1/4, animations: {
                
                self.index += 1;
                self.imgView?.transform = CGAffineTransform(rotationAngle: .pi/2*self.index)
            })
            
            UIView.addKeyframe(withRelativeStartTime: 3/4, relativeDuration: 1/4, animations: {
                
                self.index += 1;
                self.imgView?.transform = CGAffineTransform(rotationAngle: .pi/2*self.index)
            })
            
        }) { (finish) in
            
            if finish {
                self.startAnimation()
            }
        }
    }
}
